
# Velvet 1.2.10
---

A [KBase](https://kbase.us) module generated by the [KBase SDK](https://github.com/kbase/kb_sdk).


This is a KBase module that wraps the open source package "Short read de novo assembler using de Bruijn graphs"
Velvet_1.2.10

[Reference website](https://github.com/dzerbino/velvet)
[Reference manual](https://github.com/dzerbino/velvet/blob/master/Columbus_manual.pdf)

<p>[Copyright 2007, 2008 Daniel Zerbino] (zerbino@ebi.ac.uk)<br>

    This is free software; see the source for copying conditions.  There is NO warranty; 
    not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
</p>

To compile and run the Velvet App implementation, run:

    cd Velvet
    make          (required after making changes to $module_name.spec)
    kb-sdk test   (will require setting test user account credentials in test_local/test.cfg)

For more help on how to modify, register and deploy the example to KBase, see the
[KBase SDK documentation](https://github.com/kbase/kb_sdk).


===============================================================
<h3>velveth, help

velveth - simple hashing program
Version 1.2.10
</h3>

    Compilation settings:
    CATEGORIES = 57
    MAXKMERLENGTH = 127
<p>
Usage:
    ./velveth out_folder hash_length {[-file_format][-read_type][-separate|-interleaved] filename1 [filename2 ...]} {...} [options]

        -out_folder      : directory name for output files
        -hash_length     : EITHER an odd integer (if even, it will be decremented) <= 31 (if above, will be reduced) 
                OR, m,M,s where m and M are odd integers (if not, they will be decremented) with m < M <= 31
                (if above, will be reduced) and s is a step (even number). Velvet will then hash from k=m to
                k=M with a step of s
        file_name       : path to sequence file or - for standard input

- File format options:

        -fasta  -fastq  -raw    -fasta.gz       -fastq.gz       -raw.gz -sam    -bam    -fmtAuto
        (Note: -fmtAuto will detect fasta or fastq, and will try the following programs for decompression:
        gunzip, pbunzip2, bunzip2

- File layout options for paired reads (only for fasta and fastq formats):

        -interleaved    : File contains paired reads interleaved in the one file (default)
        -separate       : Read 2 separate files for paired reads

- Read type options:

        -short  -shortPaired
        -short2 -shortPaired2
        -long   -longPaired
        -reference

- Options:
        -strand_specific        : for strand specific transcriptome sequencing data (default: off)
        -reuse_Sequences        : reuse Sequences file (or link) already in directory (no need to provide 
                original filenames in this case (default: off)
        -reuse_binary   : reuse binary sequences file (or link) already in directory (no need to provide 
                original filenames in this case (default: off)
        -noHash                 : simply prepare Sequences file, do not hash reads or prepare Roadmaps 
                file (default: off)
        -create_binary          : create binary CnyUnifiedSeq file (default: off)

Synopsis:

- Short single end reads:

        velveth Assem 29 -short -fastq s_1_sequence.txt

- Paired-end short reads (remember to interleave paired reads):

        velveth Assem 31 -shortPaired -fasta interleaved.fna

- Paired-end short reads (using separate files for the paired reads)

        velveth Assem 31 -shortPaired -fasta -separate left.fa right.fa

- Two channels and some long reads:

        velveth Assem 43 -short -fastq unmapped.fna -longPaired -fasta SangerReads.fasta

- Three channels:

        velveth Assem 35 -shortPaired -fasta pe_lib1.fasta -shortPaired2 pe_lib2.fasta -short3 se_lib1.fa

Output: 

        The folder name that contains the following two subfolders/files:
        out_folder/Roadmaps
        out_folder/Sequences
                [Both files are picked up by graph, so please leave them there]
</p>
   
==============================================================
<h3>velvetg --help</h3>
usage:  

        ./velvetg wk_folder [options]
        wk_folder                       : working directory name

<p>
Standard options:

        -cov_cutoff <floating-point|auto>       : removal of low coverage nodes AFTER tour bus or allow the
                system to infer it (default: no removal)
        -ins_length <integer>           : expected distance between two paired end reads (default: no read pairing)
        -read_trkg <yes|no>             : tracking of short read positions in assembly (default: no tracking)
        -min_contig_lgth <integer>      : minimum contig length exported to contigs.fa file (default: hash length * 2)
        -amos_file <yes|no>             : export assembly to AMOS file (default: no export)
        -exp_cov <floating point|auto>  : expected coverage of unique regions or allow the system to infer it
                (default: no long or paired-end read resolution)
        -long_cov_cutoff <floating-point>: removal of nodes with low long-read coverage AFTER tour bus
                (default: no removal)

Advanced options:

        -ins_length* <integer>          : expected distance between two paired-end reads in the respective 
                short-read dataset (default: no read pairing)
        -ins_length_long <integer>      : expected distance between two long paired-end reads (default: no read pairing)
        -ins_length*_sd <integer>       : est. standard deviation of respective dataset (default: 10% of corresponding length)
                [replace '*' by nothing, '2' or '_long' as necessary]
        -scaffolding <yes|no>           : scaffolding of contigs used paired end information (default: on)
        -max_branch_length <integer>    : maximum length in base pair of bubble (default: 100)
        -max_divergence <floating-point>: maximum divergence rate between two branches in a bubble (default: 0.2)
        -max_gap_count <integer>        : maximum number of gaps allowed in the alignment of the two branches of a bubble (default: 3)
        -min_pair_count <integer>       : minimum number of paired end connections to justify the scaffolding of two long contigs (default: 5)
        -max_coverage <floating point>  : removal of high coverage nodes AFTER tour bus (default: no removal)
        -coverage_mask <int>    : minimum coverage required for confident regions of contigs (default: 1)
        -long_mult_cutoff <int>         : minimum number of long reads required to merge contigs (default: 2)
        -unused_reads <yes|no>          : export unused reads in UnusedReads.fa file (default: no)
        -alignments <yes|no>            : export a summary of contig alignment to the reference sequences (default: no)
        -exportFiltered <yes|no>        : export the long nodes which were eliminated by the coverage filters (default: no)
        -clean <yes|no>                 : remove all the intermediary files which are useless for recalculation (default : no)
        -very_clean <yes|no>            : remove all the intermediary files (no recalculation possible) (default: no)
        -paired_exp_fraction <float>   : remove all the paired end connections which less than the specified fraction of the expected count (default: 0.1)
        -shortMatePaired* <yes|no>      : for mate-pair libraries, indicate that the library might be contaminated with paired-end reads (default no)
        -conserveLong <yes|no>          : preserve sequences with long reads in them (default no)

Output:

        wk_folder/contigs.fa            : fasta file of contigs longer than twice hash length
        wk_folder/stats.txt             : stats file (tab-spaced) useful for determining appropriate coverage cutoff
        wk_folder/LastGraph             : special formatted file with all the information on the final graph
        wk_folder/velvet_asm.afg        : (if requested) AMOS compatible assembly file

    Example 1******************************:<br>
    root@a8f57e8985e8:/kb/module# /kb/module/velvet/velvetg /kb/module/work/tmp/velveth_outfolder

<br>
    root@a8f57e8985e8:/kb/module# ls -al /kb/module/work/tmp/velveth_outfolder/
        total 16652
        drwxr-xr-x 10 root root      340 May  3 15:10 .
        drwxr-xr-x  3 root root      102 May  3 14:43 ..
        -rw-r--r--  1 root root   200669 May  3 15:10 Graph
        -rw-r--r--  1 root root   200669 May  3 15:10 LastGraph
        -rw-r--r--  1 root root     1302 May  3 15:10 Log
        -rw-r--r--  1 root root   108173 May  3 15:10 PreGraph
        -rw-r--r--  1 root root  6358231 May  3 15:06 Roadmaps
        -rw-r--r--  1 root root 10063561 May  3 15:06 Sequences
        -rw-r--r--  1 root root   102424 May  3 15:10 contigs.fa
        -rw-r--r--  1 root root      912 May  3 15:10 stats.txt

    root@a8f57e8985e8:/kb/module# cat /kb/module/work/tmp/velveth_outfolder/stats.txt 
        ID      lgth    out     in      long_cov        short1_cov      short1_Ocov     short2_cov      short2_Ocov     long_nb short1_nb       short2_nb
        1       23736   2       2       0.000000        18.616153       18.597110       0.000000        0.000000        0       0       0
        2       6591    2       0       0.000000        18.814899       18.807768       0.000000        0.000000        0       0       0
        3       24184   2       2       0.000000        18.632774       18.622767       0.000000        0.000000        0       0       0
        4       414     2       2       0.000000        18.190821       18.188406       0.000000        0.000000        0       0       0
        5       44966   2       0       0.000000        18.710159       18.690277       0.000000        0.000000        0       0       0
        6       21      1       1       0.000000        15.428571       15.428571       0.000000        0.000000        0       0       0
        7       21      1       1       0.000000        1.190476        1.190476        0.000000        0.000000        0       0       0
        8       21      1       1       0.000000        18.904762       18.904762       0.000000        0.000000        0       0       0
        9       21      1       1       0.000000        14.761905       13.571429       0.000000        0.000000        0       0       0
        10      21      1       1       0.000000        22.047619       22.047619       0.000000        0.000000        0       0       0
        11      21      1       1       0.000000        1.190476        1.190476        0.000000        0.000000        0       0       0
        12      21      1       1       0.000000        1.190476        1.190476        0.000000        0.000000        0       0       0
        13      21      1       1       0.000000        1.047619        1.047619        0.000000        0.000000        0       0       0

    
    Example 2******************************:<br>
    root@a8f57e8985e8:/kb/module# /kb/module/velvet/velvetg /kb/module/work/tmp/velveth_outfolder -cov_cutoff 0.1
        [0.000000] Reading roadmap file /kb/module/work/tmp/velveth_outfolder/Roadmaps
        ....................
        [0.286045] Writing into graph file /kb/module/work/tmp/velveth_outfolder/LastGraph...
    Final graph has 13 nodes and n50 of 24184, max 44966, total 100059, using 0/142858 reads
</p>

