
# Velvet
---

A [KBase](https://kbase.us) module generated by the [KBase SDK](https://github.com/kbase/kb_sdk).


This is a KBase module that wraps the open source package "Short read de novo assembler using de Bruijn graphs"
Velvet_1.2.10

References:
https://github.com/dzerbino/velvet
https://github.com/dzerbino/velvet/blob/master/Columbus_manual.pdf

Copyright 2007, 2008 Daniel Zerbino (zerbino@ebi.ac.uk)
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

To compile and run the Velvet App implementation, run:

    cd Velvet
    make          (required after making changes to $module_name.spec)
    kb-sdk test   (will require setting test user account credentials in test_local/test.cfg)

For more help on how to modify, register and deploy the example to KBase, see the
[KBase SDK documentation](https://github.com/kbase/kb_sdk).


===============================================================
velveth, help

velveth - simple hashing program
Version 1.2.10

Compilation settings:
CATEGORIES = 2
MAXKMERLENGTH = 31

Usage:
./velveth out_folder hash_length {[-file_format][-read_type][-separate|-interleaved] filename1 [filename2 ...]} {...} [options]

        out_folder       : directory name for output files
        hash_length     : EITHER an odd integer (if even, it will be decremented) <= 31 (if above, will be reduced)
                        : OR: m,M,s where m and M are odd integers (if not, they will be decremented) with m < M <= 31 (if above, will be reduced)
                                and s is a step (even number). Velvet will then hash from k=m to k=M with a step of s
        file_name        : path to sequence file or - for standard input

File format options:
        -fasta  -fastq  -raw    -fasta.gz       -fastq.gz       -raw.gz -sam    -bam    -fmtAuto
        (Note: -fmtAuto will detect fasta or fastq, and will try the following programs for decompression : gunzip, pbunzip2, bunzip2

File layout options for paired reads (only for fasta and fastq formats):
        -interleaved    : File contains paired reads interleaved in the one file (default)
        -separate       : Read 2 separate files for paired reads

Read type options:
        -short  -shortPaired
        -short2 -shortPaired2
        -long   -longPaired
        -reference

Options:
        -strand_specific        : for strand specific transcriptome sequencing data (default: off)
        -reuse_Sequences        : reuse Sequences file (or link) already in directory (no need to provide original filenames in this case (default: off)
        -reuse_binary   : reuse binary sequences file (or link) already in directory (no need to provide original filenames in this case (default: off)
        -noHash                 : simply prepare Sequences file, do not hash reads or prepare Roadmaps file (default: off)
        -create_binary          : create binary CnyUnifiedSeq file (default: off)

Synopsis:

- Short single end reads:
        velveth Assem 29 -short -fastq s_1_sequence.txt

- Paired-end short reads (remember to interleave paired reads):
        velveth Assem 31 -shortPaired -fasta interleaved.fna

- Paired-end short reads (using separate files for the paired reads)
        velveth Assem 31 -shortPaired -fasta -separate left.fa right.fa

- Two channels and some long reads:
        velveth Assem 43 -short -fastq unmapped.fna -longPaired -fasta SangerReads.fasta

- Three channels:
        velveth Assem 35 -shortPaired -fasta pe_lib1.fasta -shortPaired2 pe_lib2.fasta -short3 se_lib1.fa

Output: The folder name that contains the following two subfolders/files:
        out_folder/Roadmaps
        out_folder/Sequences
                [Both files are picked up by graph, so please leave them there]

    Here is the test examples and their stdout printouts:
root@c50eaaa56231:/kb/module# ls /data
__READY__  velvet_data
root@c50eaaa56231:/kb/module# cd /velvet_data/
root@c50eaaa56231:/velvet_data# ls
test_long.fa  test_reads.fa  test_reads.sam  test_reference.fa
root@c50eaaa56231:/velvet_data# sort test_reads.sam > mySortedReads.sam
root@c50eaaa56231:/velvet_data# velveth test_dir 21 -reference test_reference.fa -shortPaired -sam mySortedReads.sam
[0.000000] Reading FastA file test_reference.fa;
[0.006270] 1 sequences found
[0.006299] Done
[0.006331] Reading SAM file mySortedReads.sam
[0.246146] 142858 reads found.
[0.246170] Done
[0.246172] Reference mapping counters
[0.246173] Name Read mappings
[0.246174] SEQUENCE     142858
[0.246222] Reading read set file test_dir/Sequences;
[0.259455] 142859 sequences found
[0.259559] Read 1 of length 32773, longer than limit 32767
[0.259575] You should modify recompile with the LONGSEQUENCES option (cf. manual)
        string workspace_name - the name of the workspace for input/output
        string out_folder; #folder name for output files
        int hash_length; #EITHER an odd integer (if even, it will be decremented) <= 31 (if above, will be reduced)L
        string filename; #path to sequence file or - for standard input
        string file_format; #e.g., -fasta, -fastq, -raw,-fasta.gz, -fastq.gz, -raw.gz, -sam, -bam, -fmtAuto
        string file_layout; #e.g., -interleaved or -separate 
        string read_type; #e.g., -short, -shortPaired, short2, shortPaired2, -long, -longPaired, or -reference
   
==============================================================
velvetg --help
 
        wk_folder                       : working directory name

Standard options:
        -cov_cutoff <floating-point|auto>       : removal of low coverage nodes AFTER tour bus or allow the system to infer it
                (default: no removal)
        -ins_length <integer>           : expected distance between two paired end reads (default: no read pairing)
        -read_trkg <yes|no>             : tracking of short read positions in assembly (default: no tracking)
        -min_contig_lgth <integer>      : minimum contig length exported to contigs.fa file (default: hash length * 2)
        -amos_file <yes|no>             : export assembly to AMOS file (default: no export)
        -exp_cov <floating point|auto>  : expected coverage of unique regions or allow the system to infer it
                (default: no long or paired-end read resolution)
        -long_cov_cutoff <floating-point>: removal of nodes with low long-read coverage AFTER tour bus
                (default: no removal)

Advanced options:
        -ins_length* <integer>          : expected distance between two paired-end reads in the respective short-read dataset (default: no read pairing)
        -ins_length_long <integer>      : expected distance between two long paired-end reads (default: no read pairing)
        -ins_length*_sd <integer>       : est. standard deviation of respective dataset (default: 10% of corresponding length)
                [replace '*' by nothing, '2' or '_long' as necessary]
        -scaffolding <yes|no>           : scaffolding of contigs used paired end information (default: on)
        -max_branch_length <integer>    : maximum length in base pair of bubble (default: 100)
        -max_divergence <floating-point>: maximum divergence rate between two branches in a bubble (default: 0.2)
        -max_gap_count <integer>        : maximum number of gaps allowed in the alignment of the two branches of a bubble (default: 3)
        -min_pair_count <integer>       : minimum number of paired end connections to justify the scaffolding of two long contigs (default: 5)
        -max_coverage <floating point>  : removal of high coverage nodes AFTER tour bus (default: no removal)
        -coverage_mask <int>    : minimum coverage required for confident regions of contigs (default: 1)
        -long_mult_cutoff <int>         : minimum number of long reads required to merge contigs (default: 2)
        -unused_reads <yes|no>          : export unused reads in UnusedReads.fa file (default: no)
        -alignments <yes|no>            : export a summary of contig alignment to the reference sequences (default: no)
        -exportFiltered <yes|no>        : export the long nodes which were eliminated by the coverage filters (default: no)
        -clean <yes|no>                 : remove all the intermediary files which are useless for recalculation (default : no)
        -very_clean <yes|no>            : remove all the intermediary files (no recalculation possible) (default: no)
        -paired_exp_fraction <float>   : remove all the paired end connections which less than the specified fraction of the expected count (default: 0.1)
        -shortMatePaired* <yes|no>      : for mate-pair libraries, indicate that the library might be contaminated with paired-end reads (default no)
        -conserveLong <yes|no>          : preserve sequences with long reads in them (default no)

Output:
        wk_folder/contigs.fa            : fasta file of contigs longer than twice hash length
        wk_folder/stats.txt             : stats file (tab-spaced) useful for determining appropriate coverage cutoff
        wk_folder/LastGraph             : special formatted file with all the information on the final graph
        wk_folder/velvet_asm.afg        : (if requested) AMOS compatible assembly file
    
Example: 
        ./velvetg wk_folder [options]
        string wk_folder; #folder name for files to work on and to save results
        float cov_cutoff; #removal of low coverage nodes AFTER tour bus or allow the system to infer it (default: no removal)
        int ins_length; #expected distance between two paired end reads (default: no read pairing)
        int read_trkg; # (1=yes|0=no) tracking of short read positions in assembly (default:0)
        int min_contig_length; #minimum contig length exported to contigs.fa file (default: hash length * 2)
        int amos_file; # (1=yes|0=no) #export assembly to AMOS file (default: 0)
        float exp_cov; # <floating point|auto>, expected coverage of unique regions or allow the system to infer it (default: no long or paired-end read resolution)
        float long_cov_cutoff; #removal of nodes with low long-read coverage AFTER tour bus(default: no removal)

Example 1******************************:
root@a8f57e8985e8:/kb/module# /kb/module/velvet/velvetg /kb/module/work/tmp/velveth_outfolder 
[0.000000] Reading roadmap file /kb/module/work/tmp/velveth_outfolder/Roadmaps
[0.186241] 142858 roadmaps read
[0.187244] Creating insertion markers
[0.200749] Ordering insertion markers
[0.274086] Counting preNodes
[0.286508] 106331 preNodes counted, creating them now
[0.442948] Adjusting marker info...
[0.457206] Connecting preNodes
[0.517518] Cleaning up memory
[0.518942] Done creating preGraph
[0.518956] Concatenation...
[0.541450] Renumbering preNodes
[0.541506] Initial preNode count 106331
[0.543644] Destroyed 83056 preNodes
[0.543674] Concatenation over!
[0.543676] Clipping short tips off preGraph
[0.547323] Concatenation...
[0.552777] Renumbering preNodes
[0.552836] Initial preNode count 23275
[0.553376] Destroyed 23070 preNodes
[0.553405] Concatenation over!
[0.553407] 11897 tips cut off
[0.553409] 205 nodes left
[0.559824] Writing into pregraph file /kb/module/work/tmp/velveth_outfolder/PreGraph...
[0.577028] Reading read set file /kb/module/work/tmp/velveth_outfolder/Sequences;
[0.631350] 142858 sequences found
[0.782756] Done
[0.838311] Reading pre-graph file /kb/module/work/tmp/velveth_outfolder/PreGraph
[0.839205] Graph has 205 nodes and 142858 sequences
[0.843581] Scanning pre-graph file /kb/module/work/tmp/velveth_outfolder/PreGraph for k-mers
[0.846083] 101403 kmers found
[0.851464] Sorting kmer occurence table ... 
[0.871973] Sorting done.
[0.872029] Computing acceleration table... 
[0.902730] Computing offsets... 
[0.904008] Ghost Threading through reads 0 / 142858
[0.906167]  === Ghost-Threaded in 0.002160 s
[0.906245] Threading through reads 0 / 142858
[1.258779]  === Threaded in 0.352534 s
[1.267668] Correcting graph with cutoff 0.200000
[1.267744] Determining eligible starting points
[1.267841] Done listing starting nodes
[1.267851] Initializing todo lists
[1.267862] Done with initilization
[1.267863] Activating arc lookup table
[1.267870] Done activating arc lookup table
[1.270729] Concatenation...
[1.270753] Renumbering nodes
[1.270755] Initial node count 205
[1.270757] Removed 192 null nodes
[1.270758] Concatenation over!
[1.270768] Clipping short tips off graph, drastic
[1.270770] Concatenation...
[1.270771] Renumbering nodes
[1.270772] Initial node count 13
[1.270773] Removed 0 null nodes
[1.270774] Concatenation over!
[1.270774] 13 nodes left
[1.272440] Writing into graph file /kb/module/work/tmp/velveth_outfolder/Graph...
[1.292097] WARNING: NO COVERAGE CUTOFF PROVIDED
[1.292154] Velvet will probably leave behind many detectable errors
[1.292156] See manual for instructions on how to set the coverage cutoff parameter
[1.292157] Removing contigs with coverage < -1.000000...
[1.292190] Concatenation...
[1.292201] Renumbering nodes
[1.292202] Initial node count 13
[1.292204] Removed 0 null nodes
[1.292205] Concatenation over!
[1.292206] Concatenation...
[1.292207] Renumbering nodes
[1.292208] Initial node count 13
[1.292217] Removed 0 null nodes
[1.292218] Concatenation over!
[1.292220] Clipping short tips off graph, drastic
[1.292221] Concatenation...
[1.292222] Renumbering nodes
[1.292222] Initial node count 13
[1.292223] Removed 0 null nodes
[1.292224] Concatenation over!
[1.292225] 13 nodes left
[1.292234] WARNING: NO EXPECTED COVERAGE PROVIDED
[1.292235] Velvet will be unable to resolve any repeats
[1.292237] See manual for instructions on how to set the expected coverage parameter
[1.292238] Concatenation...
[1.292239] Renumbering nodes
[1.292240] Initial node count 13
[1.292241] Removed 0 null nodes
[1.292241] Concatenation over!
[1.292242] Removing reference contigs with coverage < -1.000000...
[1.292253] Concatenation...
[1.292254] Renumbering nodes
[1.292255] Initial node count 13
[1.292256] Removed 0 null nodes
[1.292257] Concatenation over!
[1.294472] Writing contigs into /kb/module/work/tmp/velveth_outfolder/contigs.fa...
[1.309443] Writing into stats file /kb/module/work/tmp/velveth_outfolder/stats.txt...
[1.310940] Writing into graph file /kb/module/work/tmp/velveth_outfolder/LastGraph...
Final graph has 13 nodes and n50 of 24184, max 44966, total 100059, using 0/142858 reads
root@a8f57e8985e8:/kb/module# ls /kb/module/work/tmp/velveth_outfolder/
Graph  LastGraph  Log  PreGraph  Roadmaps  Sequences  contigs.fa  stats.txt
root@a8f57e8985e8:/kb/module# ls -al /kb/module/work/tmp/velveth_outfolder/
total 16652
drwxr-xr-x 10 root root      340 May  3 15:10 .
drwxr-xr-x  3 root root      102 May  3 14:43 ..
-rw-r--r--  1 root root   200669 May  3 15:10 Graph
-rw-r--r--  1 root root   200669 May  3 15:10 LastGraph
-rw-r--r--  1 root root     1302 May  3 15:10 Log
-rw-r--r--  1 root root   108173 May  3 15:10 PreGraph
-rw-r--r--  1 root root  6358231 May  3 15:06 Roadmaps
-rw-r--r--  1 root root 10063561 May  3 15:06 Sequences
-rw-r--r--  1 root root   102424 May  3 15:10 contigs.fa
-rw-r--r--  1 root root      912 May  3 15:10 stats.txt
root@a8f57e8985e8:/kb/module# cat /kb/module/work/tmp/velveth_outfolder/stats.txt 
ID      lgth    out     in      long_cov        short1_cov      short1_Ocov     short2_cov      short2_Ocov     long_nb short1_nb       short2_nb
1       23736   2       2       0.000000        18.616153       18.597110       0.000000        0.000000        0       0       0
2       6591    2       0       0.000000        18.814899       18.807768       0.000000        0.000000        0       0       0
3       24184   2       2       0.000000        18.632774       18.622767       0.000000        0.000000        0       0       0
4       414     2       2       0.000000        18.190821       18.188406       0.000000        0.000000        0       0       0
5       44966   2       0       0.000000        18.710159       18.690277       0.000000        0.000000        0       0       0
6       21      1       1       0.000000        15.428571       15.428571       0.000000        0.000000        0       0       0
7       21      1       1       0.000000        1.190476        1.190476        0.000000        0.000000        0       0       0
8       21      1       1       0.000000        18.904762       18.904762       0.000000        0.000000        0       0       0
9       21      1       1       0.000000        14.761905       13.571429       0.000000        0.000000        0       0       0
10      21      1       1       0.000000        22.047619       22.047619       0.000000        0.000000        0       0       0
11      21      1       1       0.000000        1.190476        1.190476        0.000000        0.000000        0       0       0
12      21      1       1       0.000000        1.190476        1.190476        0.000000        0.000000        0       0       0
13      21      1       1       0.000000        1.047619        1.047619        0.000000        0.000000        0       0       0


Example 2******************************:
root@a8f57e8985e8:/kb/module# /kb/module/velvet/velvetg /kb/module/work/tmp/velveth_outfolder -cov_cutoff 0.1

[0.000000] Reading graph file /kb/module/work/tmp/velveth_outfolder/Graph
[0.000648] Graph has 13 nodes and 142858 sequences
[0.007546] Reading read set file /kb/module/work/tmp/velveth_outfolder/Sequences;
[0.051878] 142858 sequences found
[0.201670] Done
[0.254594] Removing contigs with coverage < 0.100000...
[0.254647] Concatenation...
[0.254650] Renumbering nodes
[0.254652] Initial node count 13
[0.254660] Removed 0 null nodes
[0.254661] Concatenation over!
[0.254663] Concatenation...
[0.254664] Renumbering nodes
[0.254666] Initial node count 13
[0.254667] Removed 0 null nodes
[0.254668] Concatenation over!
[0.254670] Clipping short tips off graph, drastic
[0.254672] Concatenation...
[0.254673] Renumbering nodes
[0.254675] Initial node count 13
[0.254676] Removed 0 null nodes
[0.254677] Concatenation over!
[0.254678] 13 nodes left
[0.254680] WARNING: NO EXPECTED COVERAGE PROVIDED
[0.254681] Velvet will be unable to resolve any repeats
[0.254683] See manual for instructions on how to set the expected coverage parameter
[0.254685] Concatenation...
[0.254686] Renumbering nodes
[0.254687] Initial node count 13
[0.254688] Removed 0 null nodes
[0.254689] Concatenation over!
[0.254690] Removing reference contigs with coverage < 0.100000...
[0.254703] Concatenation...
[0.254708] Renumbering nodes
[0.254712] Initial node count 13
[0.254717] Removed 0 null nodes
[0.254721] Concatenation over!
[0.259348] Writing contigs into /kb/module/work/tmp/velveth_outfolder/contigs.fa...
[0.282974] Writing into stats file /kb/module/work/tmp/velveth_outfolder/stats.txt...
[0.286045] Writing into graph file /kb/module/work/tmp/velveth_outfolder/LastGraph...
Final graph has 13 nodes and n50 of 24184, max 44966, total 100059, using 0/142858 reads

